### PR TITLE
Issue 635

### LINKED ISSUE #635: spring-cloud-kubernetes-config options and autoconfigure not working with reload.
**Describe the bug**

Attempting to configure  spring-cloud-kubernetes-config to just add support for kubernetes configMaps. (and disable secrets and discovery client) However would like to support config reload with event. It appears that when bringing up spring boot with the configuration below along with AutoConfigure, we run into a  bean resolution issue as posted below.

**Sample Configuration**

```
spring:
  cloud:
    kubernetes:
      config:
        enabled: true
      secrets:
        enabled: false
      discovery:
        enabled: false
      reload:
        enabled: true
        monitoring-config-maps: true
```

Exception:

```
Caused by: org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'org.springframework.cloud.kubernetes.config.reload.ConfigReloadAutoConfiguration$ConfigReloadAutoConfigurationBeans': Unsatisfied dependency expressed through field 'secretsPropertySourceLocator'; nested exception is org.springframework.beans.factory.NoSuchBeanDefinitionException: No qualifying bean of type 'org.springframework.cloud.kubernetes.config.SecretsPropertySourceLocator' available: expected at least 1 bean which qualifies as autowire candidate. Dependency annotations: {@org.springframework.beans.factory.annotation.Autowired(required=true)}
```

If the above seems a valid set of configs, Ideally, i would expect the configuration flags that either define or not define a given bean to be honored for autoconfigs as well.

Tested with:

```
Spring Boot: 2.3.3.RELEASE
spring-cloud-kubernetes-config: 1.1.3.RELEASE
```

### PR DESCRIPTION
**Pull request by krisiye**
Fixes #635 

1. Makes Autowired dependencies under ConfigReloadAutoConfiguration and ConfigReloadDefaultAutoConfiguration optional. That way it is consistent with conditionals on BootstrapConfiguration. The ConfigurationChangeDetector/watcher downstream honors configs for configMap or secret and registers a watcher as expected.
2. Adds tests for use-cases where either config and/or secret is enabled or vice-versa

### CODE REVIEW

File: spring-cloud-kubernetes-config/src/main/java/org/springframework/cloud/kubernetes/config/reload/ConfigReloadAutoConfiguration.java [hunk 1]
```diff
@@ -71,10 +71,10 @@
 		@Autowired
 		private KubernetesClient kubernetesClient;
 
-		@Autowired
+		@Autowired(required = false)
 		private ConfigMapPropertySourceLocator configMapPropertySourceLocator;
```
[the remark this answers is not in the document]
Comment (krisiye) [reply spring-cloud-kubernetes-config/src/main/java/org/springframework/cloud/kubernetes/config/reload/ConfigReloadAutoConfiguration.java:75]: @ryanjbaxter  - Thanks for the review! The ConfigurationChangeDetector implementation takes care of the conditions and does appropriate checking before the watchers (polling/event) are registered. I believe we should be okay with that? Also I can resubmit this against 1.1.x.  Do you want me to close this in favor of that? Just FYI - this issue also exists on the master branch.

[the remark this answers is not in the document]
Comment (krisiye) [reply spring-cloud-kubernetes-config/src/main/java/org/springframework/cloud/kubernetes/config/reload/ConfigReloadAutoConfiguration.java:75]: Submitted #639 against 1.1.x

[the remark this answers is not in the document]
Comment (ryanjbaxter) [reply spring-cloud-kubernetes-config/src/main/java/org/springframework/cloud/kubernetes/config/reload/ConfigReloadAutoConfiguration.java:75]: @krisiye are you talking about here?

https://github.com/spring-cloud/spring-cloud-kubernetes/blob/master/spring-cloud-kubernetes-config/src/main/java/org/springframework/cloud/kubernetes/config/reload/ConfigurationChangeDetector.java#L180

There is no null check...

Besides the preferred pattern here would to not autowire them and just inject them into the method.

[the remark this answers is not in the document]
Comment (krisiye) [reply spring-cloud-kubernetes-config/src/main/java/org/springframework/cloud/kubernetes/config/reload/ConfigReloadAutoConfiguration.java:75]: @ryanjbaxter  - I meant the property checks here - https://github.com/spring-cloud/spring-cloud-kubernetes/blob/master/spring-cloud-kubernetes-config/src/main/java/org/springframework/cloud/kubernetes/config/reload/EventBasedConfigurationChangeDetector.java#L68 . Sure we could probably get rid of Autowired dependency here to make it a bit more cleaner.

[the remark this answers is not in the document]
Comment (ryanjbaxter) [reply spring-cloud-kubernetes-config/src/main/java/org/springframework/cloud/kubernetes/config/reload/ConfigReloadAutoConfiguration.java:75]: @krisiye but how does that prevent `configMapPropertySourceLocator` from being null and potentially causing a NPE?

[the remark this answers is not in the document]
Comment (krisiye) [reply spring-cloud-kubernetes-config/src/main/java/org/springframework/cloud/kubernetes/config/reload/ConfigReloadAutoConfiguration.java:75]: @ryanjbaxter  - Looking into this a bit more, my understanding is that Bootstrap is responsible for (https://github.com/spring-cloud/spring-cloud-kubernetes/blob/master/spring-cloud-kubernetes-config/src/main/java/org/springframework/cloud/kubernetes/config/BootstrapConfiguration.java#L50-L60) conditionally defining the PropertySourceLocator beans (both configMap and Secrets).  Injection into the ChangeDetecter either by autowired (required=true) or method is not going to work. Any suggestions?

[the remark this answers is not in the document]
Comment (ryanjbaxter) [reply spring-cloud-kubernetes-config/src/main/java/org/springframework/cloud/kubernetes/config/reload/ConfigReloadAutoConfiguration.java:75]: In other words is a`ConfigurationChangeDetector` requires a `PropertySourceLocator` whats the point of creating one if it is null?

[the remark this answers is not in the document]
Comment (ryanjbaxter) [reply spring-cloud-kubernetes-config/src/main/java/org/springframework/cloud/kubernetes/config/reload/ConfigReloadAutoConfiguration.java:75]: @krisiye that is why I am saying to create it conditionally

[the remark this answers is not in the document]
Comment (krisiye) [reply spring-cloud-kubernetes-config/src/main/java/org/springframework/cloud/kubernetes/config/reload/ConfigReloadAutoConfiguration.java:75]: @ryanjbaxter  - correct me if i am wrong, you are suggesting we create ConfigurationChangeDetector beans conditionally based on either properties configMap/secrets are enabled and in addition we would add null checks under the polling/event change detectors. Right?

[the remark this answers is not in the document]
Comment (ryanjbaxter) [reply spring-cloud-kubernetes-config/src/main/java/org/springframework/cloud/kubernetes/config/reload/ConfigReloadAutoConfiguration.java:76]: @krisiye No what I am thinking is adding `@ConditionalOnBean({ ConfigMapPropertySourceLocator.class, SecretsPropertySourceLocator.class })` on `propertyChangeWatcher`

[the remark this answers is not in the document]
Comment (krisiye) [reply spring-cloud-kubernetes-config/src/main/java/org/springframework/cloud/kubernetes/config/reload/ConfigReloadAutoConfiguration.java:76]: @ryanjbaxter  okay to close this PR and make these updates on #639  ?

[the remark this answers is not in the document]
Comment (krisiye) [reply spring-cloud-kubernetes-config/src/main/java/org/springframework/cloud/kubernetes/config/reload/ConfigReloadAutoConfiguration.java:76]: @ryanjbaxter  - Updated #639 with review comments. thanks!

**Pull request closed without merging**